discovery-provider/src/tasks/index_tastemaker_notifications.py

```diff
@@ -0,0 +1,112 @@
+from datetime import datetime
+from typing import List
+
+from sqlalchemy import asc
+from src.models.notifications.notification import Notification
+from src.models.social.repost import Repost
+from src.models.social.save import Save
+from src.models.tracks.track import Track
+from src.utils.session_manager import SessionManager
+
+
+def create_tastemaker_group_id(user_id, repost_item_id):
+    return f"tastemaker:{user_id}:tastemaker_item_id:{repost_item_id}"
+
+
+def index_tastemaker_notifications(
+    db: SessionManager,
+    top_trending_tracks: List[Track],
+    tastemaker_notification_threshold=10,
+):
+    with db.scoped_session() as session:
+        tastemaker_notifications = []
+        for track in top_trending_tracks:
+            (
+                repost_tastemaker_notifications,
+                existing_group_ids,
+            ) = create_action_tastemaker_notifications(
+                tastemaker_notification_threshold,
+                session,
+                existing_group_ids=set(),
+                track=track,
+                action_type=Repost,
+            )
+            tastemaker_notifications.extend(repost_tastemaker_notifications)
+
+            (
+                save_tastemaker_notifications,
+                existing_group_ids,
+            ) = create_action_tastemaker_notifications(
+                tastemaker_notification_threshold,
+                session,
+                existing_group_ids=existing_group_ids,
+                track=track,
+                action_type=Save,
+            )
+            tastemaker_notifications.extend(save_tastemaker_notifications)
+
+        session.bulk_save_objects(
+            [notification for notification in tastemaker_notifications]
+        )
+
+
+def create_action_tastemaker_notifications(
+    tastemaker_notification_threshold,
+    session,
+    existing_group_ids,
+    track,
+    action_type,
+):
+    query_action_item_id = (
```

Comment:
can we combine this and the "action_item_id" below?

Reply:
also, i wonder if it's possible to update the query to check both repost or saved? so we can just do one query instead of two? basically is there a way to generate all the relevant notifs in one shot?

Reply:
ugh i dont think so.. one is referencing `Repost.repost_item_id` to be used in the query and the other is referencing `repost.repost_item_id` as a value like `1` or something

i cant think of a way to query all at once... cuz we have to query two different tables to get the earliest reposts and earliest saves. this was my best attempt at dry-ing up the code since its the same query for both but i agree its a little awkward cuz of their diff attribute names 😬